Mathlib/CategoryTheory/Monoidal/Coherence.lean

```diff
@@ -0,0 +1,393 @@
+  [MonoidalCoherence W (X ⊗ (Y ⊗ Z))] : MonoidalCoherence W ((X ⊗ Y) ⊗ Z) :=
+⟨MonoidalCoherence.hom ≫ (α_ X Y Z).inv⟩
+
+end MonoidalCoherence
+
+/-- Construct an isomorphism between two objects in a monoidal category
+out of unitors and associators. -/
+def monoidalIso (X Y : C) [LiftObj X] [LiftObj Y] [MonoidalCoherence X Y] : X ≅ Y :=
+asIso MonoidalCoherence.hom
+
+example (X : C) : X ≅ (X ⊗ (𝟙_ C ⊗ 𝟙_ C)) := monoidalIso _ _
+
+example (X1 X2 X3 X4 X5 X6 X7 X8 X9 : C) :
+  (𝟙_ C ⊗ (X1 ⊗ X2 ⊗ ((X3 ⊗ X4) ⊗ X5)) ⊗ X6 ⊗ (X7 ⊗ X8 ⊗ X9)) ≅
+  (X1 ⊗ (X2 ⊗ X3) ⊗ X4 ⊗ (X5 ⊗ (𝟙_ C ⊗ X6) ⊗ X7) ⊗ X8 ⊗ X9) :=
+monoidalIso _ _
+
+/-- Compose two morphisms in a monoidal category,
+inserting unitors and associators between as necessary. -/
+def monoidalComp {W X Y Z : C} [LiftObj X] [LiftObj Y]
+  [MonoidalCoherence X Y] (f : W ⟶ X) (g : Y ⟶ Z) : W ⟶ Z :=
+f ≫ MonoidalCoherence.hom ≫ g
+
+@[inherit_doc monoidalComp]
+infixr:80 " ⊗≫ " => monoidalComp -- type as \ot \gg
+
+/-- Compose two isomorphisms in a monoidal category,
+inserting unitors and associators between as necessary. -/
+def monoidalIsoComp {W X Y Z : C} [LiftObj X] [LiftObj Y]
+  [MonoidalCoherence X Y] (f : W ≅ X) (g : Y ≅ Z) : W ≅ Z :=
+f ≪≫ asIso MonoidalCoherence.hom ≪≫ g
+
+@[inherit_doc monoidalIsoComp]
+infixr:80 " ≪⊗≫ " => monoidalIsoComp -- type as \ot \gg
+
+example {U V W X Y : C} (f : U ⟶ V ⊗ (W ⊗ X)) (g : (V ⊗ W) ⊗ X ⟶ Y) : U ⟶ Y := f ⊗≫ g
+
+-- To automatically insert unitors/associators at the beginning or end,
+-- you can use `f ⊗≫ 𝟙 _`
+example {W X Y Z : C} (f : W ⟶ (X ⊗ Y) ⊗ Z) : W ⟶ X ⊗ (Y ⊗ Z) := f ⊗≫ 𝟙 _
+
+@[simp] lemma monoidalComp_refl {X Y Z : C} (f : X ⟶ Y) (g : Y ⟶ Z) :
+    f ⊗≫ g = f ≫ g := by
+  simp [monoidalComp]
+
+example {U V W X Y : C} (f : U ⟶ V ⊗ (W ⊗ X)) (g : (V ⊗ W) ⊗ X ⟶ Y) :
+    f ⊗≫ g = f ≫ (α_ _ _ _).inv ≫ g := by
+  simp [monoidalComp]
+
+end CategoryTheory.MonoidalCategory
+
+open CategoryTheory.MonoidalCategory
+
+namespace Mathlib.Tactic.Coherence
+
+open Lean Meta Elab Tactic
+
+/--
+Auxilliary definition of `monoidal_coherence`,
+being careful with namespaces to avoid shadowing.
+-/
+-- We could construct this expression directly without using `elabTerm`,
+-- but it would require preparing many implicit arguments by hand.
+def mkProjectMapExpr (e : Expr) : TermElabM Expr := do
+  Term.elabTerm (← `(CategoryTheory.FreeMonoidalCategory.projectMap _root_.id _ _
+    (CategoryTheory.MonoidalCategory.LiftHom.lift $(← Term.exprToSyntax e)))) none
+
+/-- Coherence tactic for monoidal categories. -/
+def monoidal_coherence (g : MVarId) : TermElabM Unit := do
+  withOptions (fun opts => opts.setNat `synthInstance.maxSize
+    (max 256 (opts.getNat `synthInstance.maxSize))) do
+  -- TODO: is this `dsimp only` step necessary? It doesn't appear to be in the tests below.
+  let (ty, _) ← dsimp (← g.getType) (← Simp.Context.ofNames [] true)
+  let some (_, lhs, rhs) := ty.eq? | throwError "Not an equation of morphisms."
```

Comment:
I think the canonical way of throwing an error is `throwTacticEx `monoidal_coherence g m!"foobar"`

Reply:
Thanks, I've adopted this.